custom-targets/vertex-ai/model-deployer/render.go

```diff
@@ -0,0 +1,339 @@
+	}, nil
+}
+
+func (r *renderer) addCommonMetadata(rs *clouddeploy.RenderResult) {
+	if rs.Metadata == nil {
+		rs.Metadata = map[string]string{}
+	}
+	rs.Metadata[clouddeploy.CustomTargetSourceMetadataKey] = aiDeployerSampleName
+	rs.Metadata[clouddeploy.CustomTargetSourceSHAMetadataKey] = clouddeploy.GitCommit
+}
+
+func validateRequest(modelNameFromDeployParameter, endpointName string, minReplicaCountParameter int64, deployedModel *aiplatform.GoogleCloudAiplatformV1DeployedModel) error {
+	modelRegion, err := fetchRegionFromModel(modelNameFromDeployParameter)
+	if err != nil {
+		return fmt.Errorf("unable to parse region from model: %v", err)
+	}
+
+	endpointRegion, err := fetchRegionFromEndpoint(endpointName)
+	if err != nil {
+		return fmt.Errorf("unable to parse region from endpoint: %v", err)
+	}
+
+	if endpointRegion != modelRegion {
+		return fmt.Errorf("The model to be deployed must be in the same region as the endpoint. Copy the model to the region the  endpoint is located, or make an endpoint in the same region as the model")
+	}
+
+	if err = verifyModelNameNotDefinedInConfig(deployedModel); err != nil {
+		return err
+	}
+
+	if err = verifyMinReplicaCountHasNoConflicts(deployedModel, minReplicaCountParameter); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func verifyMinReplicaCountHasNoConflicts(deployedModel *aiplatform.GoogleCloudAiplatformV1DeployedModel, deployParameterValue int64) error {
+
+	configValue := minReplicaCountFromConfig(deployedModel)
+
+	// checks if minReplicaCount is not defined either in deploy parameter or config file
+	if configValue == deployParameterValue {
+		if configValue == 0 {
+			return fmt.Errorf("minReplicaCount must either be defined in the config file or provided to the render operation through a deploy parameter using 'vertexAIMinReplicaCount' key")
+		}
+	}
+
+	// only other valid format is if either but not both are 0
+	if configValue == 0 || deployParameterValue == 0 {
+		return nil
+	}
+	return fmt.Errorf("the minReplicaCount parameter is defined in both the provided config file and as a deploy parameter and both values differ from each other, please define minReplicaCount in the config file or as a deploy-parameter")
+}
+
+func minReplicaCountFromConfig(deployedModel *aiplatform.GoogleCloudAiplatformV1DeployedModel) int64 {
+	if deployedModel.DedicatedResources != nil {
+		return deployedModel.DedicatedResources.MinReplicaCount
+	}
+	return 0
+}
+func verifyModelNameNotDefinedInConfig(deployedModel *aiplatform.GoogleCloudAiplatformV1DeployedModel) error {
+
+	if deployedModel.Model != "" {
+		return fmt.Errorf("model to deployed must be supplied as a deploy parameter and not in the config file")
+	}
+
+	if deployedModel.ModelVersionId != "" {
+		return fmt.Errorf("the model version id to deploy must be supplied as part of the vertexAIModel deployparamater containing the model to be deployed")
+	}
+
+	return nil
+
+}
+
+// renderDeployModelRequest generates a model definition
```

Comment:
More detailed comment, remember this is a sample that's meant to be digestible by others. Same applies to the other functions.